modules/templates/templates/templates.test.tsx

```diff
@@ -0,0 +1,732 @@
+    });
+
+    it("Checks template can be edited", async () => {
+      render(<Templates />);
+
+      // Wait for the table to be fully rendered (at least one row is filled)
+      await waitFor(() => {
+        expect(screen.getByTestId("table-row-0"));
+      });
+
+      const templateEdit = screen.getAllByTestId("templates.edit");
+      act(() => templateEdit[0].click());
+
+      const inputElement = screen.getByTestId("templates.name-edit-field");
+      await userEvent.clear(inputElement);
+      await userEvent.type(inputElement, "New Template Name");
+
+      expect(inputElement).toHaveValue("New Template Name");
+
+      const saveButton = screen.getByTestId("templates.save-template");
+      expect(saveButton).toBeInTheDocument();
+
+      act(() => saveButton.click());
+
+      await waitFor(() => {
+        expect(
+          screen.queryByTestId("templates.name-edit-field"),
+        ).not.toBeInTheDocument();
+      });
+
+      expect(screen.getByText("New Template Name"));
+    });
+
+    it("Checks modal pops up when delete icon clicked", async () => {
+      render(<Templates />);
+
+      // Wait for the table to be fully rendered (at least one row is filled)
+      await waitFor(() => {
+        expect(screen.getByTestId("table-row-0"));
+      });
+
+      const templateEdit = screen.getAllByTestId("templates.delete");
+      act(() => templateEdit[0].click());
+
+      await waitFor(() => {
+        expect(
+          screen.getByTestId("templates.delete-confirmation"),
+        ).toBeInTheDocument();
+      });
+    });
+
+    it("Checks template delete function", async () => {
+      render(<Templates />);
+
+      await waitFor(() => {
+        expect(screen.getByTestId("table-row-0"));
+      });
+
+      const templateEdit = screen.getAllByTestId("templates.delete");
+      act(() => templateEdit[0].click());
+
+      await waitFor(() => {
+        expect(
+          screen.getByTestId("templates.delete-confirmation"),
+        ).toBeInTheDocument();
+      });
+
+      const confirmButton = screen.getByTestId("modal.confirm-button");
+      expect(confirmButton).toBeInTheDocument();
+      act(() => confirmButton.click());
+
+      await waitFor(() => {
+        expect(
+          screen.queryByTestId("templates.delete-confirmation"),
+        ).not.toBeInTheDocument();
+      });
+
+      // TODO: Make this check dynamic
+      expect(screen.queryByText("New Template Name")).not.toBeInTheDocument();
+    });
+
+    it("Checks template redirect on select", async () => {
+      const spy = jest.spyOn(RouterMock, "push");
+      render(<Templates />);
+
+      await waitFor(() => {
+        expect(screen.getByTestId("table-row-0"));
+      });
+
+      const headers = screen.getAllByTestId("templates.title-cell");
+      act(() => headers[1].click());
+
+      expect(spy).toHaveBeenCalledTimes(1);
+      expect(spy).toHaveBeenCalledWith(
+        expect.stringMatching(/\/app\/template-detail\?id=.*/),
+      );
```

Comment:
May no need to create a spay for this test cases, can test without spy directly using RouterMock (I learn this at the end)